real_time/surface_src/java_src/Alice/src/main/java/com/roboclub/robobuggy/nodes/localizers/KfLocalizer.java

```diff
@@ -1,43 +1,308 @@
+								{ gpsUTM.getNorthing() }, { 0 }, { 0 }, { th },
+								{ 0 }, { 0 } };
+						double[][] updateCovariance = {
+								{ 1, 0, 0, 0, 0, 0, 0 }, // x
+								{ 0, 1, 0, 0, 0, 0, 0 }, // y
+								{ 0, 0, 1, 0, 0, 0, 0 }, // x_dot
+								{ 0, 0, 0, 1, 0, 0, 0 }, // y_dot
+								{ 0, 0, 0, 0, 1, 0, 0 }, // th
+								{ 0, 0, 0, 0, 0, 1, 0 }, // th_dot
+								{ 0, 0, 0, 0, 0, 0, 1 } // heading
+						};
+						updateStep(new Matrix(observationModel), new Matrix(
+								meassurement), new Matrix(updateCovariance));
+					}
+				});
+
+		/*
+		new Subscriber(
+				"HighTrustGpsLoc",
+				NodeChannel.IMU_ANG_POS.getMsgPath(),
+				((topicName, m) -> {
+
+					IMUAngularPositionMessage mes = ((IMUAngularPositionMessage) m);
+					// double y = mes.getRot()[0][1];
+					// double x = mes.getRot()[0][0];
+					double[][] xVar = { { 1 }, { 0 }, { 0 } };
+					double[][] yVar = { { 0 }, { 1 }, { 0 } };
+					Matrix xMat = new Matrix(xVar);
+					Matrix yMat = new Matrix(yVar);
+					Matrix rot = new Matrix(mes.getRot());
+					double x = rot.times(xMat).get(0, 0);
+					double y = rot.times(yMat).get(0, 0);
+					double th = -(Math.toDegrees(Math.atan2(y, x)) - 90);
+					double[][] observationModel = { { 0, 0, 0, 0, 0, 0, 0 }, // x
+							{ 0, 0, 0, 0, 0, 0, 0 }, // y
+							{ 0, 0, 0, 0, 0, 0, 0 }, // x_dot_b
+							{ 0, 0, 0, 0, 0, 0, 0 }, // y_dot_b
+							{ 0, 0, 0, 0, 1, 0, 0 }, // th
+							{ 0, 0, 0, 0, 0, 0, 0 }, // th_dot
+							{ 0, 0, 0, 0, 0, 0, 0 } // Heading
+					};
+
+					double[][] meassurement = { { 0 }, { 0 }, { 0 }, { 0 },
+							{ th }, { 0 }, { 0 } };
+					double[][] updateCovariance = { { 1, 0, 0, 0, 0, 0, 0 }, // x
+							{ 0, 1, 0, 0, 0, 0, 0 }, // y
+							{ 0, 0, 1, 0, 0, 0, 0 }, // x_dot
+							{ 0, 0, 0, 1, 0, 0, 0 }, // y_dot
+							{ 0, 0, 0, 0, 1, 0, 0 }, // th
+							{ 0, 0, 0, 0, 0, 1, 0 }, // th_dot
+							{ 0, 0, 0, 0, 0, 0, 1 } // heading
+					};
+					// updateStep(new Matrix(observationModel),new
+					// Matrix(meassurement),new Matrix(updateCovariance));
+
+				}));
+				*/
+
+		// TODO note that we will probably run into precision errors since the
+		// changes are so small
+		// would be good to batch up the encoder updates until we get a margin
+		// that we know can be represented proeprly
+		new Subscriber("htGpsLoc", NodeChannel.ENCODER.getMsgPath(),
+				new MessageListener() {
+					@Override
+					public synchronized void actionPerformed(String topicName,
+							Message m) {
+						EncoderMeasurement measurement = (EncoderMeasurement) m;
+
+						// convert the feet from the last message into a delta
+						// degree, and update our position
+						double currentEncoderMeasurement = measurement
+								.getDistance();
+						double deltaDistance = currentEncoderMeasurement
+								- lastEncoderReading;
+						long currentTime = new Date().getTime();
+						long dt = currentTime - lastEncoderReadingTime;
+						if (dt > 1) { // to remove numeric instability
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid using Literals in Conditional Statements](https://www.codacy.com/app/Mr-Trevor-Decker/RoboBuggy/file/2684047960/issues/source?bid=2938665&fileBranchId=3246677#l199)